test/bench.ml

```diff
@@ -0,0 +1,100 @@
+let time_fn ~f =
+  let t = Unix.gettimeofday () in
+  let res = f () in
+  let time_elapsed = Unix.gettimeofday () -. t in
+  (res, time_elapsed)
+
+type result = {
+  bench_name : string;
+  median_exec_time : float;
+  avg_exec_time : float;
+  num_of_runs : int;
+  shortest_exec_time : float;
+  longest_exec_time : float;
+}
+
+let run ?(pre = Fun.id) ?(post = ignore) ?(runs = 10) ~name ~f () =
+  assert (runs > 0);
+  let runtimes =
+    Array.init runs (fun _ ->
+        pre ();
+        let f_res, time_elapsed = time_fn ~f in
+        let () = post f_res in
+        time_elapsed)
+  in
+  let () = Array.sort Float.compare runtimes in
+  let runtime_sum = Array.fold_left (fun acc x -> acc +. x) 0. runtimes in
+  let median =
+    begin
+      if runs mod 2 = 0 then begin
+        (runtimes.((runs + 1) / 2) +. runtimes.((runs - 1) / 2)) /. 2.0
+      end
+      else runtimes.((runs + 1) / 2)
+    end
+  in
+  {
+    bench_name = name;
+    median_exec_time = median;
+    avg_exec_time = runtime_sum /. Float.of_int runs;
+    num_of_runs = runs;
+    shortest_exec_time = runtimes.(0);
+    longest_exec_time = runtimes.(runs - 1);
+  }
+
+let process s =
+  let first_n_after_nonzero n s =
+    let non_zero_at = ref (-1) in
+    let () =
+      String.iteri
+        (fun i c -> if c <> '0' && !non_zero_at = -1 then non_zero_at := i)
+        s
+    in
+    String.sub s 0 (min (!non_zero_at + n) (String.length s))
+  in
+
+  let l = String.split_on_char '.' s in
+  match l with
+  | [x] -> x
+  | [before; after] -> before ^ "." ^ first_n_after_nonzero 4 after
+  | _ -> failwith "Impossible"
+
+let report ?(in' = `Ms) res =
+  let name_padding =
+    List.fold_left
+      (fun acc x -> Int.max acc (String.length x.bench_name))
+      10 res
+  in
+  let pad_with_space x len =
+    assert (String.length x <= len);
+    x ^ String.init (len - String.length x) (fun _ -> ' ')
+  in
+  let f = pad_with_space in
+  let time_unit_s =
+    match in' with `S -> "s" | `Ms -> "ms" | `Us -> "µs" | `Ns -> "ns"
+  in
+  let g x =
+    let y =
+      match in' with
+      | `S -> x
+      | `Ms -> x *. 1000.
+      | `Us -> x *. 1000. *. 1000.
+      | `Ns -> x *. 1000. *. 1000. *. 1000.
```

Comment:
BTW, this doesn't make a difference here and you might already know this, but floating point operations are generally not associative so the order matters for the compiler.  If you have `x *. 1000. *. 1000. *. 1000.` the compiler more or less needs to generate 3 multiply instructions.  If you reorder to `1000. *. 1000. *. 1000. *. x` or parenthesize to `x *. (1000. *. 1000. *. 1000.)` then the compiler can constant fold part of the computation and generate just a single multiply instruction.  See [here](https://ocaml.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:ocaml,selection:(endColumn:1,endLineNumber:14,positionColumn:1,positionLineNumber:14,selectionStartColumn:1,selectionStartLineNumber:14,startColumn:1,startLineNumber:14),source:'let+fast+in!'+x+%3D%0A++++++match+in!'+with%0A++++++%7C+%60S+-%3E+x%0A++++++%7C+%60Ms+-%3E+1000.+*.+x%0A++++++%7C+%60Us+-%3E+1000.+*.+1000.+*.+x%0A++++++%7C+%60Ns+-%3E+1000.+*.+1000.+*.+1000.+*.+x%0A%0Alet+slow+in!'+x+%3D%0A++++++match+in!'+with%0A++++++%7C+%60S+-%3E+x%0A++++++%7C+%60Ms+-%3E+x+*.+1000.%0A++++++%7C+%60Us+-%3E+x+*.+1000.+*.+1000.%0A++++++%7C+%60Ns+-%3E+x+*.+1000.+*.+1000.+*.+1000.%0A'),l:'5',n:'0',o:'OCaml+source+%231',t:'0')),k:50.00730169680667,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:ocaml5000,deviceViewOpen:'1',filters:(b:'1',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:ocaml,libs:!(),options:'',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+ocamlopt+5.0.0+(Editor+%231)',t:'0')),k:49.99269830319334,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

Reply:
Aah yes. I'll change it.